src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,146 @@
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->tableName, 'm')
+                ->where('m.delivered_at is null')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->driverConnection->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL()
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $this->driverConnection->createQueryBuilder()
+                ->update($this->tableName)
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id')
+                ->setParameter('delivered_at', (new \DateTime('now'))->format('Y-m-d H:i:s'))
+                ->setParameter('id', $doctrineEnvelope['id'])
+                ->execute();
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Exception $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->tableName, ['id' => $id]) > 0;
+    }
+
+    public function nack($id): bool
+    {
+        return $this->driverConnection->update($this->tableName, ['delivered_at' => null], ['id' => $id]) > 0;
+    }
+
+    public function requeueMessages()
+    {
+        $redeliverLimit = new \DateTime('now');
+        $redeliverLimit->modify(sprintf('-%d second', $this->redeliverTimeout));
+        $this->driverConnection->createQueryBuilder()
+            ->update($this->tableName)
+            ->set('delivered_at', ':delivered_at')
+            ->andWhere('delivered_at < :redeliver_limit')
+            ->setParameter('delivered_at', null)
+            ->setParameter('redeliver_limit', $redeliverLimit->format('Y-m-d H:i:s'))
+            ->execute();
+    }
+
+    private function setup(): void
+    {
+        $synchronizer = new SingleDatabaseSynchronizer($this->driverConnection);
+        $synchronizer->updateSchema($this->getSchema(), true);
+    }
+
+    private function getSchema(): Schema
+    {
+        $schema = new Schema();
+        $table = $schema->createTable($this->tableName);
+        $table->addColumn('id', Type::BIGINT)
+            ->setAutoincrement(true)
+            ->setNotnull(true);
+        $table->addColumn('body', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('headers', Type::TEXT)
```

Comment:
Why not using `JSON`?

Reply:
I use the `text` type because I thought that `JSON` was not implemented in all DBMS supported by Doctrine DBAL.

But I see that MySQL support the JSON type since the 5.7 version. I didn't check ALL DBMS but maybe we can use the `JSON` type :thinking:

Reply:
Doesn't this depend on the serializer? The `PhpSerializer` doesn't use `headers` but in theory, it could, and it would be a `serialize()` string - not JSON.

Reply:
For the body correct, not for the headers. Though, let's keep it consistent and keep `text` 👍